### PR TITLE
feat: adding unit tests for `DetrayGeometryConverter` and `DetrayMaterialConverter`

### PR DESCRIPTION
**Pull request by asalzburger**
This PR adds unit tests for the new detray converters.

Nota bene, that only the conversion into the `detray::payload` layer is tested here, more excessive validation will be done using the `Examples` and an according detray testing suite.

it also fixes a bug in the `DetrayJsonConverter` which was found by comparing to the direct conversion.

### CODE REVIEW

**Comment by asalzburger** [comment]
Hey @EleniXoch - I've put you as a reviewer, because this could be a good guideline for your surface grid converter to also come up with one or two unit tests :-P

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  // Translate a cylinder
+  auto cylinderSurface = Acts::Surface::makeShared<CylinderSurface>(
+      Transform3::Identity(), std::make_shared<CylinderBounds>(20., 100.));
+
+  detray::io::surface_payload payload = DetrayGeometryConverter::convertSurface(
+      tContext, *cylinderSurface, false);
+
+  // Check the payload
+  BOOST_CHECK(!payload.index_in_coll.has_value());
+  BOOST_CHECK(payload.mask.shape == detray::io::shape_id::cylinder2);
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:153]: Can you check that the source link has been set? Or does the cylinder here not have a GeoId, yet?

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+}
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:154]: Also you could add ```BOOST_CHECK(payload.type != detray::surface_id::e_portal)``` (if the ctlinder geoID does not have a random value)

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+  // Convert the detector
+  DetrayConversionUtils::GeometryIdCache geoIdCache;
+
+  detray::io::detector_payload payload =
+      DetrayGeometryConverter::convertDetector(geoIdCache, tContext, *detector,
+                                               *logger);
+
+  // test the payload
+  BOOST_CHECK_EQUAL(payload.volumes.size(), 6u);
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:262]: Will this get more test cases? That seems to be a lot of setup for only one test case

File: Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+auto logger =
+    Acts::getDefaultLogger("DetrayMaterialConverterTests", Acts::Logging::INFO);
+
+BOOST_AUTO_TEST_CASE(DetrayMaterialSlabConversion) {
+  // Convert the material slab
+  detray::io::material_slab_payload payload =
+      Acts::DetrayMaterialConverter::convertMaterialSlab(materialSlab12345);
+
+  // Material tuype should be set to slab
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp:47]:
```suggestion
  // Material type should be set to slab
```

File: Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+                           detray::io::material_id>
+      payload = Acts::DetrayMaterialConverter::convertSurfaceMaterial(
+          binnedMaterial, *logger);
+  // Check the payload
+  BOOST_CHECK_EQUAL(payload.axes.size(), 2u);
+  BOOST_CHECK(payload.grid_link.type ==
+              detray::io::material_id::rectangle2_map);
+  //  The axis are real x-y
+  BOOST_CHECK(payload.axes[0u].label == detray::axis::label::e_x);
+  BOOST_CHECK(payload.axes[1u].label == detray::axis::label::e_y);
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp:174]: Can we also check the number of bins for 2x2 binned material?

File: Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+  BOOST_CHECK_EQUAL(payload.axes[0u].bins, 4u);
+  CHECK_CLOSE_ABS(payload.axes[0u].edges[0u], -2.,
+                  std::numeric_limits<Acts::ActsScalar>::epsilon());
+  CHECK_CLOSE_ABS(payload.axes[0u].edges[1u], 2.,
+                  std::numeric_limits<Acts::ActsScalar>::epsilon());
+  BOOST_CHECK(payload.axes[0u].binning == detray::axis::binning::e_regular);
+  BOOST_CHECK(payload.axes[0u].bounds == detray::axis::bounds::e_closed);
+  // axis is dummy
+  BOOST_CHECK(payload.axes[1u].label == detray::axis::label::e_y);
+  BOOST_CHECK_EQUAL(payload.axes[1u].bins, 1u);
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp:107]: Will you test somewhere that the local bin indices are assigned correctly to the slab payloads?

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp (hunk 1 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:153]: Ok, will update to set geo ids and check for the source link.

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp (hunk 3 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:262]: Actually,  yes, particularly the portal splitting test is something I want to do.

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp (hunk 1 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:153]: Added.

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp (hunk 2 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:154]: Added.

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp (hunk 3 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:262]: Added all the portal splitting tests, looks good!

File: Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp (hunk 6 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp:107]: Let's do it.

File: Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp (hunk 5 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Detray/DetrayMaterialConverterTests.cpp:218]: Done.

File: Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+  cylinderSurface->assignGeometryId(sgID);
+
+  detray::io::surface_payload payload = DetrayGeometryConverter::convertSurface(
+      tContext, *cylinderSurface, false);
+
+  // Check the payload
+  BOOST_CHECK(!payload.index_in_coll.has_value());
+  BOOST_CHECK(payload.mask.shape == detray::io::shape_id::cylinder2);
+  BOOST_CHECK_EQUAL(payload.source, sgID.value());
+  BOOST_CHECK(payload.type != detray::surface_id::e_portal);
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Detray/DetrayGeometryConverterTests.cpp:158]: If the cylinder can be set to sensitive now, then we can make the check a bit tighter:
```suggestion
  BOOST_CHECK(payload.type == detray::surface_id::e_sensitive);
```

**Review by niermann999 — COMMENTED** [review]
One tiny comment, otherwise looks good!

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [5084be542c1820478b79e920e9bc83b38bd67f0d]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922892)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922891)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922890)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922889)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922888)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922887)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922886)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922885)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922884)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922883)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922882)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922881)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922880)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922879)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922878)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922877)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922876)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922875)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922874)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922873)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922872)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922871)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922870)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922869)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922868)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922867)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922866)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922865)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922864)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922863)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922862)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922861)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42922860)||

**Pull request merged**